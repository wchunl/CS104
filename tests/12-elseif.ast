ROOT "" (0.0.0)
|   TOK_FUNCTION "" (6.22.1) function
|   |   TOK_TYPE_ID "" (6.22.1) typeid
|   |   |   TOK_VOID "void" (6.22.1) void
|   |   |   TOK_IDENT "fail" (6.22.6) variable
|   |   TOK_PARAM "(" (6.22.11) param
|   |   |   TOK_TYPE_ID "" (6.22.12) typeid
|   |   |   |   TOK_STRING "string" (6.22.12) string
|   |   |   |   TOK_IDENT "expr" (6.22.19) string variable (6.22.12)
|   |   |   TOK_TYPE_ID "" (6.22.25) typeid
|   |   |   |   TOK_STRING "string" (6.22.25) string
|   |   |   |   TOK_IDENT "file" (6.22.32) string variable (6.22.25)
|   |   |   TOK_TYPE_ID "" (6.22.38) typeid
|   |   |   |   TOK_INT "int" (6.22.38) int
|   |   |   |   TOK_IDENT "line" (6.22.42) int variable (6.22.38)
|   TOK_FUNCTION "" (6.24.1) function
|   |   TOK_TYPE_ID "" (6.24.1) typeid
|   |   |   TOK_VOID "void" (6.24.1) void
|   |   |   TOK_IDENT "putchr" (6.24.6) variable
|   |   TOK_PARAM "(" (6.24.13) param
|   |   |   TOK_TYPE_ID "" (6.24.14) typeid
|   |   |   |   TOK_INT "int" (6.24.14) int
|   |   |   |   TOK_IDENT "chr" (6.24.18) int variable (6.24.14)
|   TOK_FUNCTION "" (6.25.1) function
|   |   TOK_TYPE_ID "" (6.25.1) typeid
|   |   |   TOK_VOID "void" (6.25.1) void
|   |   |   TOK_IDENT "putint" (6.25.6) variable
|   |   TOK_PARAM "(" (6.25.13) param
|   |   |   TOK_TYPE_ID "" (6.25.14) typeid
|   |   |   |   TOK_INT "int" (6.25.14) int
|   |   |   |   TOK_IDENT "num" (6.25.18) int variable (6.25.14)
|   TOK_FUNCTION "" (6.26.1) function
|   |   TOK_TYPE_ID "" (6.26.1) typeid
|   |   |   TOK_VOID "void" (6.26.1) void
|   |   |   TOK_IDENT "putstr" (6.26.6) variable
|   |   TOK_PARAM "(" (6.26.13) param
|   |   |   TOK_TYPE_ID "" (6.26.14) typeid
|   |   |   |   TOK_STRING "string" (6.26.14) string
|   |   |   |   TOK_IDENT "str" (6.26.21) string variable (6.26.14)
|   TOK_FUNCTION "" (6.28.1) function
|   |   TOK_TYPE_ID "" (6.28.1) typeid
|   |   |   TOK_INT "int" (6.28.1) int
|   |   |   TOK_IDENT "getchr" (6.28.5) variable
|   |   TOK_PARAM "(" (6.28.11) param
|   TOK_FUNCTION "" (6.29.1) function
|   |   TOK_TYPE_ID "" (6.29.1) typeid
|   |   |   TOK_STRING "string" (6.29.1) string
|   |   |   TOK_IDENT "getstr" (6.29.8) variable
|   |   TOK_PARAM "(" (6.29.14) param
|   TOK_FUNCTION "" (6.30.1) function
|   |   TOK_TYPE_ID "" (6.30.1) typeid
|   |   |   TOK_STRING "string" (6.30.1) string
|   |   |   TOK_IDENT "getln" (6.30.8) variable
|   |   TOK_PARAM "(" (6.30.13) param
|   TOK_FUNCTION "" (7.5.1) function
|   |   TOK_TYPE_ID "" (7.5.1) typeid
|   |   |   TOK_INT "int" (7.5.1) int
|   |   |   TOK_IDENT "main" (7.5.5) variable
|   |   TOK_PARAM "(" (7.5.10) param
|   |   |   TOK_TYPE_ID "" (7.5.11) typeid
|   |   |   |   TOK_INT "int" (7.5.11) int
|   |   |   |   TOK_IDENT "argc" (7.5.15) int variable (7.5.11)
|   |   TOK_BLOCK "{" (7.5.21)
|   |   |   TOK_IF "if" (7.6.4)
|   |   |   |   TOK_EQ "==" (7.6.13) int vreg
|   |   |   |   |   TOK_IDENT "argc" (7.6.8) int variable (7.5.11)
|   |   |   |   |   TOK_INTCON "1" (7.6.16) int const
|   |   |   |   TOK_CALL "(" (7.6.26)
|   |   |   |   |   TOK_IDENT "putstr" (7.6.19) variable
|   |   |   |   |   TOK_STRINGCON ""one"" (7.6.27) string const
|   |   |   |   TOK_IFELSE "if" (7.7.9)
|   |   |   |   |   TOK_EQ "==" (7.7.18) int vreg
|   |   |   |   |   |   TOK_IDENT "argc" (7.7.13) int variable (7.5.11)
|   |   |   |   |   |   TOK_INTCON "2" (7.7.21) int const
|   |   |   |   |   TOK_CALL "(" (7.7.31)
|   |   |   |   |   |   TOK_IDENT "putstr" (7.7.24) variable
|   |   |   |   |   |   TOK_STRINGCON ""two"" (7.7.32) string const
|   |   |   |   |   TOK_IFELSE "if" (7.8.9)
|   |   |   |   |   |   TOK_EQ "==" (7.8.18) int vreg
|   |   |   |   |   |   |   TOK_IDENT "argc" (7.8.13) int variable (7.5.11)
|   |   |   |   |   |   |   TOK_INTCON "3" (7.8.21) int const
|   |   |   |   |   |   TOK_CALL "(" (7.8.31)
|   |   |   |   |   |   |   TOK_IDENT "putstr" (7.8.24) variable
|   |   |   |   |   |   |   TOK_STRINGCON ""three"" (7.8.32) string const
|   |   |   |   |   |   TOK_IFELSE "(" (7.9.16)
|   |   |   |   |   |   |   TOK_IDENT "putstr" (7.9.9) variable
|   |   |   |   |   |   |   TOK_STRINGCON ""many"" (7.9.17) string const
|   |   |   TOK_CALL "(" (7.10.11)
|   |   |   |   TOK_IDENT "putchr" (7.10.4) variable
|   |   |   |   TOK_CHARCON "'\n'" (7.10.12) int const
|   |   |   TOK_RETURN "return" (7.11.4)
|   |   |   |   TOK_INTCON "0" (7.11.11) int const
