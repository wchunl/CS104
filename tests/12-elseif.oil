fail:     .function
          .param ptr expr
          .param ptr file
          .param int line
          .end
putchr:   .function
          .param int chr
          .end
putint:   .function
          .param int num
          .end
putstr:   .function
          .param ptr str
          .end
getchr:   .function int
          .end
getstr:   .function
          .end
getln:    .function
          .end
main:     .function int
          .param int argc
.if0:     $t0:i = argc == 1
          goto .el0 if not $t0:i
.th0:     putstr("one")
.el0:     
.if1:     $t1:i = argc == 2
          goto .el1 if not $t1:i
.th1:     putstr("two")
.el1:     
.if2:     $t2:i = argc == 3
          goto .el2 if not $t2:i
.th2:     putstr("three")
.el2:     putstr("many")
.fi2:     
.fi1:     
.fi0:     putchr('\n')
          return 0
          return 
          .end
