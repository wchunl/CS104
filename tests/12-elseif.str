string_set[   0]:    4706206694200543916 0x1a3adc8->","
                   16579916199511856337 0x1a3c488->"getln"
string_set[   1]:    2297668033614959926 0x1a3de88->"0"
                    3701972582333163920 0x1a3c998->"argc"
string_set[   3]:    8248777770799913213 0x1a3d668->"3"
                   11868120863286193964 0x1a3c6f8->"main"
                    9611428142580219387 0x1a36758->"fail"
string_set[   4]:      21606623943142546 0x1a3b278->"{"
                    4017852018407011111 0x1a3aed8->"file"
string_set[   6]:   13686721134187162920 0x1a3de18->"}"
string_set[   8]:    9386170080689500517 0x1a3ac78->"expr"
string_set[  12]:   15816506932648033945 0x1a3aae8->"("
string_set[  14]:    9317282556992397945 0x1a3ccb8->"if"
string_set[  17]:   15712226061040878727 0x1a3cf88->""one""
                   18413295556186398326 0x1a3b2c8->"putint"
string_set[  18]:    7211130878751977441 0x1a3b418->"putchr"
                   12679825168387489523 0x1a3b9a8->"num"
string_set[  23]:   11166701219533395393 0x1a3c1f8->"getstr"
                   12868445110721718657 0x1a3aef8->")"
string_set[  28]:   14635384904416252381 0x1a3dd38->"return"
string_set[  29]:   12291392672316406109 0x1a3b0b8->"line"
string_set[  30]:    9509744370891421448 0x1a3db08->""many""
string_set[  31]:    6192799629375248351 0x1a3dcf8->"'\n'"
                   11780556996745140861 0x1a3d918->""three""
                   14736352883882276958 0x1a3ce28->"=="
                    6142509188972423790 0x1a36738->""
                   13348923631108115385 0x1a3b738->"putstr"
string_set[  33]:   13084547564353593782 0x1a3d478->""two""
                    1583576014053361350 0x1a3bb58->"str"
string_set[  38]:   12597418120160561823 0x1a3acf8->"int"
string_set[  39]:    1873634190020599143 0x1a3d178->"else"
                    7296690267450279874 0x1a3b588->"chr"
string_set[  40]:   11911796782241914524 0x1a36318->"void"
string_set[  42]:    5344662657899890615 0x1a3b0d8->";"
string_set[  43]:    4551451650890805270 0x1a3d3c8->"2"
string_set[  44]:   13497578857207247412 0x1a3ab98->"string"
string_set[  45]:   10159970873491820195 0x1a3ced8->"1"
string_set[  46]:    9201800968376702500 0x1a3bf88->"getchr"
load_factor = 0.809
bucket_count = 47
max_bucket_size = 5
