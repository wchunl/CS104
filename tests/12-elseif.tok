#  1 "12-elseif.oc"
#  1 "<built-in>"
#  1 "<command-line>"
#  1 "12-elseif.oc"
#  1 "oclib.h"
# 22 "oclib.h"
   6 22.00001 261 TOK_VOID      void
   6 22.00006 282 TOK_IDENT     fail
   6 22.00011  40 '('           (
   6 22.00012 263 TOK_STRING    string
   6 22.00019 282 TOK_IDENT     expr
   6 22.00023  44 ','           ,
   6 22.00025 263 TOK_STRING    string
   6 22.00032 282 TOK_IDENT     file
   6 22.00036  44 ','           ,
   6 22.00038 262 TOK_INT       int
   6 22.00042 282 TOK_IDENT     line
   6 22.00046  41 ')'           )
   6 22.00047  59 ';'           ;
   6 24.00001 261 TOK_VOID      void
   6 24.00006 282 TOK_IDENT     putchr
   6 24.00013  40 '('           (
   6 24.00014 262 TOK_INT       int
   6 24.00018 282 TOK_IDENT     chr
   6 24.00021  41 ')'           )
   6 24.00022  59 ';'           ;
   6 25.00001 261 TOK_VOID      void
   6 25.00006 282 TOK_IDENT     putint
   6 25.00013  40 '('           (
   6 25.00014 262 TOK_INT       int
   6 25.00018 282 TOK_IDENT     num
   6 25.00021  41 ')'           )
   6 25.00022  59 ';'           ;
   6 26.00001 261 TOK_VOID      void
   6 26.00006 282 TOK_IDENT     putstr
   6 26.00013  40 '('           (
   6 26.00014 263 TOK_STRING    string
   6 26.00021 282 TOK_IDENT     str
   6 26.00024  41 ')'           )
   6 26.00025  59 ';'           ;
   6 28.00001 262 TOK_INT       int
   6 28.00005 282 TOK_IDENT     getchr
   6 28.00011  40 '('           (
   6 28.00012  41 ')'           )
   6 28.00013  59 ';'           ;
   6 29.00001 263 TOK_STRING    string
   6 29.00008 282 TOK_IDENT     getstr
   6 29.00014  40 '('           (
   6 29.00015  41 ')'           )
   6 29.00016  59 ';'           ;
   6 30.00001 263 TOK_STRING    string
   6 30.00008 282 TOK_IDENT     getln
   6 30.00013  40 '('           (
   6 30.00014  41 ')'           )
   6 30.00015  59 ';'           ;
#  4 "12-elseif.oc"
   7 5.00001 262 TOK_INT       int
   7 5.00005 282 TOK_IDENT     main
   7 5.00010  40 '('           (
   7 5.00011 262 TOK_INT       int
   7 5.00015 282 TOK_IDENT     argc
   7 5.00019  41 ')'           )
   7 5.00021 123 '{'           {
   7 6.00004 265 TOK_IF        if
   7 6.00007  40 '('           (
   7 6.00008 282 TOK_IDENT     argc
   7 6.00013 275 TOK_EQ        ==
   7 6.00016 283 TOK_INTCON    1
   7 6.00017  41 ')'           )
   7 6.00019 282 TOK_IDENT     putstr
   7 6.00026  40 '('           (
   7 6.00027 285 TOK_STRINGCON "one"
   7 6.00032  41 ')'           )
   7 6.00033  59 ';'           ;
   7 7.00004 266 TOK_ELSE      else
   7 7.00009 265 TOK_IF        if
   7 7.00012  40 '('           (
   7 7.00013 282 TOK_IDENT     argc
   7 7.00018 275 TOK_EQ        ==
   7 7.00021 283 TOK_INTCON    2
   7 7.00022  41 ')'           )
   7 7.00024 282 TOK_IDENT     putstr
   7 7.00031  40 '('           (
   7 7.00032 285 TOK_STRINGCON "two"
   7 7.00037  41 ')'           )
   7 7.00038  59 ';'           ;
   7 8.00004 266 TOK_ELSE      else
   7 8.00009 265 TOK_IF        if
   7 8.00012  40 '('           (
   7 8.00013 282 TOK_IDENT     argc
   7 8.00018 275 TOK_EQ        ==
   7 8.00021 283 TOK_INTCON    3
   7 8.00022  41 ')'           )
   7 8.00024 282 TOK_IDENT     putstr
   7 8.00031  40 '('           (
   7 8.00032 285 TOK_STRINGCON "three"
   7 8.00039  41 ')'           )
   7 8.00040  59 ';'           ;
   7 9.00004 266 TOK_ELSE      else
   7 9.00009 282 TOK_IDENT     putstr
   7 9.00016  40 '('           (
   7 9.00017 285 TOK_STRINGCON "many"
   7 9.00023  41 ')'           )
   7 9.00024  59 ';'           ;
   7 10.00004 282 TOK_IDENT     putchr
   7 10.00011  40 '('           (
   7 10.00012 284 TOK_CHARCON   '\n'
   7 10.00016  41 ')'           )
   7 10.00017  59 ';'           ;
   7 11.00004 268 TOK_RETURN    return
   7 11.00011 283 TOK_INTCON    0
   7 11.00012  59 ';'           ;
   7 12.00001 125 '}'           }
