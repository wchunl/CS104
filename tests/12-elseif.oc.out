; 0.  1: # 1 "12-elseif.oc"; 1.  0: 
; 1.  1: # 1 "<built-in>"
; 2.  1: # 1 "<command-line>"
; 3.  1: # 1 "12-elseif.oc"
; 4.  1: 
; 4.  2: 
; 4.  3: # 1 "oclib.h" 1
; 5.  1: # 22 "oclib.h"
; 6. 22: void fail (string expr, string file, int line);
; 6. 23: 
; 6. 24: void putchr (int chr);
; 6. 25: void putint (int num);
; 6. 26: void putstr (string str);
; 6. 27: 
; 6. 28: int getchr();
; 6. 29: string getstr();
; 6. 30: string getln();
; 6. 31: # 4 "12-elseif.oc" 2
; 7.  4: 
; 7.  5: int main (int argc) {
; 7.  6:    if (argc == 1) putstr ("one");
; 7.  7:    else if (argc == 2) putstr ("two");
; 7.  8:    else if (argc == 3) putstr ("three");
; 7.  9:    else putstr ("many");
; 7. 10:    putchr ('\n');
; 7. 11:    return 0;
; 7. 12: }
